In [1]:
import pandas as pd

In [2]:
headers = pd.read_csv("Desktop/Data Science/DBD/mnheaders.csv")

In [3]:
raw = pd.read_csv("Desktop/Data Science/DBD/mn.csv")

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (110,111,112,120) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [5]:
colnames = list(raw.columns)

# Renaming column headers

In [6]:
print(colnames)

['Unnamed: 0', 'HH1', 'HH2', 'LN', 'MWM1', 'MWM2', 'MWM4', 'MWM5', 'MWM6D', 'MWM6M', 'MWM6Y', 'MWM7', 'MWM8', 'MWM9', 'MWM10H', 'MWM10M', 'MWM11H', 'MWM11M', 'MWB1M', 'MWB1Y', 'MWB2', 'MWB3', 'MWB4', 'MWB5', 'MWB7', 'MMT2', 'MMT3', 'MMT4', 'MMT6', 'MMT7', 'MMT8', 'MMT9', 'MMT10', 'MMT11', 'MMT12', 'MMT13', 'MMT14', 'MCM1', 'MCM3', 'MCM4', 'MCM5A', 'MCM5B', 'MCM6', 'MCM7A', 'MCM7B', 'MCM8', 'MCM9A', 'MCM9B', 'MCM10', 'MCM11A', 'MCM11B', 'MCM12M', 'MCM12Y', 'MDV1A', 'MDV1B', 'MDV1C', 'MDV1D', 'MDV1E', 'MDV1F', 'MMA1', 'MMA3', 'MMA4', 'MMA5', 'MMA6', 'MMA7', 'MMA8M', 'MMA8Y', 'MMA9', 'MSB1', 'MSB2', 'MSB3U', 'MSB3N', 'MSB4', 'MSB5', 'MSB8', 'MSB9', 'MSB10', 'MSB13', 'MSB14', 'MSB15', 'MHA1', 'MHA2', 'MHA3', 'MHA4', 'MHA5', 'MHA6', 'MHA7', 'MHA8A', 'MHA8B', 'MHA8C', 'MHA9', 'MHA10', 'MHA11', 'MHA12', 'MHA24', 'MHA25', 'MHA26', 'MHA27', 'MMC1', 'MMC2', 'MMC3', 'MMC4', 'MTA1', 'MTA2', 'MTA3', 'MTA4', 'MTA5', 'MTA6', 'MTA7', 'MTA8A', 'MTA8B', 'MTA8C', 'MTA8D', 'MTA8E', 'MTA8X', 'MTA9', 'MTA10

In [7]:
len(colnames)

159

In [8]:
header_name = list(headers["Name"])

In [9]:
print(header_name)

['HH1', 'HH2', 'LN', 'MWM1', 'MWM2', 'MWM4', 'MWM5', 'MWM6D', 'MWM6M', 'MWM6Y', 'MWM7', 'MWM8', 'MWM9', 'MWM10H', 'MWM10M', 'MWM11H', 'MWM11M', 'MWB1M', 'MWB1Y', 'MWB2', 'MWB3', 'MWB4', 'MWB5', 'MWB7', 'MMT2', 'MMT3', 'MMT4', 'MMT6', 'MMT7', 'MMT8', 'MMT9', 'MMT10', 'MMT11', 'MMT12', 'MMT13', 'MMT14', 'MCM1', 'MCM3', 'MCM4', 'MCM5A', 'MCM5B', 'MCM6', 'MCM7A', 'MCM7B', 'MCM8', 'MCM9A', 'MCM9B', 'MCM10', 'MCM11A', 'MCM11B', 'MCM12D', 'MCM12M', 'MCM12Y', 'MDV1A', 'MDV1B', 'MDV1C', 'MDV1D', 'MDV1E', 'MMA1', 'MMA2', 'MMA3', 'MMA4', 'MMA5', 'MMA6', 'MMA7', 'MMA8M', 'MMA8Y', 'MMA9', 'MSB1', 'MSB2', 'MSB3U', 'MSB3N', 'MSB4', 'MSB5', 'MSB7', 'MSB8', 'MSB9', 'MSB10', 'MSB12', 'MSB13', 'MSB14', 'MSB15', 'MHA1', 'MHA2', 'MHA3', 'MHA4', 'MHA5', 'MHA6', 'MHA7', 'MHA8A', 'MHA8B', 'MHA8C', 'MHA9', 'MHA10', 'MHA11', 'MHA12', 'MHA24', 'MHA25', 'MHA26', 'MHA27', 'MTA1', 'MTA2', 'MTA3', 'MTA4', 'MTA5', 'MTA6', 'MTA7', 'MTA8A', 'MTA8B', 'MTA8C', 'MTA8D', 'MTA8X', 'MTA9', 'MTA10', 'MTA11', 'MTA12A', 'MTA12B

In [10]:
len(header_name)

209

In [11]:
# Forming a dictionary to match the current headers to a more descriptive name
table = pd.Series(headers.Label.values,index=headers.Name).to_dict()

In [12]:
table

{'HH1': 'Cluster number',
 'HH2': 'Household number',
 'LN': 'Line number',
 'MWM1': 'Cluster number',
 'MWM2': 'Household number',
 'MWM4': "Man's line number",
 'MWM5': 'Interviewer number',
 'MWM6D': 'Day of interview',
 'MWM6M': 'Month of interview',
 'MWM6Y': 'Year of interview',
 'MWM7': "Result of man's interview",
 'MWM8': 'Field editor',
 'MWM9': 'Data entry clerk',
 'MWM10H': 'Start of interview - Hour',
 'MWM10M': 'Start of interview - Minutes',
 'MWM11H': 'End of interview - Hour',
 'MWM11M': 'End of interview - Minutes',
 'MWB1M': 'Month of birth of man',
 'MWB1Y': 'Year of birth of man',
 'MWB2': 'Age of man',
 'MWB3': 'Ever attended school',
 'MWB4': 'Highest level of school attended',
 'MWB5': 'Highest grade completed at that level',
 'MWB7': 'Can read part of the sentence',
 'MMT2': 'Frequency of reading newspaper or magazine',
 'MMT3': 'Frequency of listening to the radio',
 'MMT4': 'Frequency of watching TV',
 'MMT6': 'Ever used a computer',
 'MMT7': 'Computer usage 

In [13]:
raw.rename(columns=table,inplace=True)

In [14]:
raw.head()

Unnamed: 0  Cluster number  Household number  Line number  Cluster number  \
0           1               1                17            1               1   
1           2               1                20            1               1   
2           3               2                 1            1               2   
3           4               2                 1            5               2   
4           5               2                 1            8               2   

   Household number  Man's line number  Interviewer number  Day of interview  \
0                17                  1                  14                 7   
1                20                  1                  14                 7   
2                 1                  1                   9                 8   
3                 1                  5                   9                12   
4                 1                  8                   9                 8   

   Month of interview  Year of interview Result of man's interview  \
0                   4               2014                 Completed   
1                   4               2014                 Completed   
2                   4               2014                 Completed   
3                   4               2014               Not at home   
4                   4               2014                 Completed   

   Field editor  Data entry clerk  Start of interview - Hour  \
0             2                20                       17.0   
1             2                20                       17.0   
2             1                40                       10.0   
3             1                40                        NaN   
4             1                40                       10.0   

   Start of interview - Minutes  End of interview - Hour  \
0                          59.0                     18.0   
1                          32.0                     17.0   
2                          37.0                     10.0   
3                           NaN                      NaN   
4                          53.0                     11.0   

   End of interview - Minutes  Month of birth of man  Year of birth of man  \
0                         7.0                    5.0                1984.0   
1                        42.0                    5.0                1976.0   
2                        52.0                    2.0                1973.0   
3                         NaN                    NaN                   NaN   
4                        10.0                    2.0                1993.0   

   Age of man Ever attended school Highest level of school attended  \
0        29.0                  Yes                           Higher   
1        37.0                  Yes                           Higher   
2        41.0                  Yes                          Primary   
3         NaN                  NaN                              NaN   
4        21.0                  Yes                        Secondary   

   Highest grade completed at that level Can read part of the sentence  \
0                                   31.0                           NaN   
1                                   31.0                           NaN   
2                                   17.0   Able to read whole sentence   
3                                    NaN                           NaN   
4                                   24.0                           NaN   

  Frequency of reading newspaper or magazine  \
0                           Almost every day   
1                       At least once a week   
2                                 Not at all   
3                                        NaN   
4                      Less than once a week   

  Frequency of listening to the radio Frequency of watching TV  \
0                At least once a week    Less than once a week   
1                          Not at all         Almost every day   
2                    Almost every day    Less 

In [15]:
# Remove the first column
raw = raw.drop(raw.columns[0],axis=1)

In [16]:
raw.head()

Cluster number  Household number  Line number  Cluster number  \
0               1                17            1               1   
1               1                20            1               1   
2               2                 1            1               2   
3               2                 1            5               2   
4               2                 1            8               2   

   Household number  Man's line number  Interviewer number  Day of interview  \
0                17                  1                  14                 7   
1                20                  1                  14                 7   
2                 1                  1                   9                 8   
3                 1                  5                   9                12   
4                 1                  8                   9                 8   

   Month of interview  Year of interview Result of man's interview  \
0                   4               2014                 Completed   
1                   4               2014                 Completed   
2                   4               2014                 Completed   
3                   4               2014               Not at home   
4                   4               2014                 Completed   

   Field editor  Data entry clerk  Start of interview - Hour  \
0             2                20                       17.0   
1             2                20                       17.0   
2             1                40                       10.0   
3             1                40                        NaN   
4             1                40                       10.0   

   Start of interview - Minutes  End of interview - Hour  \
0                          59.0                     18.0   
1                          32.0                     17.0   
2                          37.0                     10.0   
3                           NaN                      NaN   
4                          53.0                     11.0   

   End of interview - Minutes  Month of birth of man  Year of birth of man  \
0                         7.0                    5.0                1984.0   
1                        42.0                    5.0                1976.0   
2                        52.0                    2.0                1973.0   
3                         NaN                    NaN                   NaN   
4                        10.0                    2.0                1993.0   

   Age of man Ever attended school Highest level of school attended  \
0        29.0                  Yes                           Higher   
1        37.0                  Yes                           Higher   
2        41.0                  Yes                          Primary   
3         NaN                  NaN                              NaN   
4        21.0                  Yes                        Secondary   

   Highest grade completed at that level Can read part of the sentence  \
0                                   31.0                           NaN   
1                                   31.0                           NaN   
2                                   17.0   Able to read whole sentence   
3                                    NaN                           NaN   
4                                   24.0                           NaN   

  Frequency of reading newspaper or magazine  \
0                           Almost every day   
1                       At least once a week   
2                                 Not at all   
3                                        NaN   
4                      Less than once a week   

  Frequency of listening to the radio Frequency of watching TV  \
0                At least once a week    Less than once a week   
1                          Not at all         Almost every day   
2                    Almost every day    Less than once a week   
3                                 NaN               

In [17]:
list(raw.columns)

['Cluster number',
 'Household number',
 'Line number',
 'Cluster number',
 'Household number',
 "Man's line number",
 'Interviewer number',
 'Day of interview',
 'Month of interview',
 'Year of interview',
 "Result of man's interview",
 'Field editor',
 'Data entry clerk',
 'Start of interview - Hour',
 'Start of interview - Minutes',
 'End of interview - Hour',
 'End of interview - Minutes',
 'Month of birth of man',
 'Year of birth of man',
 'Age of man',
 'Ever attended school',
 'Highest level of school attended',
 'Highest grade completed at that level',
 'Can read part of the sentence',
 'Frequency of reading newspaper or magazine',
 'Frequency of listening to the radio',
 'Frequency of watching TV',
 'Ever used a computer',
 'Computer usage in the last 12 months',
 'Frequency of computer usage in the last month',
 'Ever used internet',
 'Internet usage in the last 12 months',
 'Frequency of Internet usage in the past month',
 'Ever used a mobile or non-mobile phone',
 'Mobile o

In [18]:
len(raw.columns)

158

Drop columns of unmatched names as we cannot be sure what they mean even after investigating the mnheaders.csv file

In [19]:
raw = raw.drop(["MDV1F", "MTA8E","mwelevel", "mnweight", "wscoreu", "windex5u", "wscorer", "windex5r"], axis = 1)

In [20]:
len(list(raw.columns))

150

In [21]:
raw.head()

Cluster number  Household number  Line number  Cluster number  \
0               1                17            1               1   
1               1                20            1               1   
2               2                 1            1               2   
3               2                 1            5               2   
4               2                 1            8               2   

   Household number  Man's line number  Interviewer number  Day of interview  \
0                17                  1                  14                 7   
1                20                  1                  14                 7   
2                 1                  1                   9                 8   
3                 1                  5                   9                12   
4                 1                  8                   9                 8   

   Month of interview  Year of interview Result of man's interview  \
0                   4               2014                 Completed   
1                   4               2014                 Completed   
2                   4               2014                 Completed   
3                   4               2014               Not at home   
4                   4               2014                 Completed   

   Field editor  Data entry clerk  Start of interview - Hour  \
0             2                20                       17.0   
1             2                20                       17.0   
2             1                40                       10.0   
3             1                40                        NaN   
4             1                40                       10.0   

   Start of interview - Minutes  End of interview - Hour  \
0                          59.0                     18.0   
1                          32.0                     17.0   
2                          37.0                     10.0   
3                           NaN                      NaN   
4                          53.0                     11.0   

   End of interview - Minutes  Month of birth of man  Year of birth of man  \
0                         7.0                    5.0                1984.0   
1                        42.0                    5.0                1976.0   
2                        52.0                    2.0                1973.0   
3                         NaN                    NaN                   NaN   
4                        10.0                    2.0                1993.0   

   Age of man Ever attended school Highest level of school attended  \
0        29.0                  Yes                           Higher   
1        37.0                  Yes                           Higher   
2        41.0                  Yes                          Primary   
3         NaN                  NaN                              NaN   
4        21.0                  Yes                        Secondary   

   Highest grade completed at that level Can read part of the sentence  \
0                                   31.0                           NaN   
1                                   31.0                           NaN   
2                                   17.0   Able to read whole sentence   
3                                    NaN                           NaN   
4                                   24.0                           NaN   

  Frequency of reading newspaper or magazine  \
0                           Almost every day   
1                       At least once a week   
2                                 Not at all   
3                                        NaN   
4                      Less than once a week   

  Frequency of listening to the radio Frequency of watching TV  \
0                At least once a week    Less than once a week   
1                          Not at all         Almost every day   
2                    Almost every day    Less than once a week   
3                                 NaN               

# Explore missing data

In [22]:
# Find out number of missing values in each column

missing = raw.isna().sum()
missing = missing.to_frame().reset_index()
missing = missing.rename(columns= {'index': 'col', 0: 'missing vals'})
missing

col  missing vals
0                                       Cluster number             0
1                                     Household number             0
2                                          Line number             0
3                                       Cluster number             0
4                                     Household number             0
5                                    Man's line number             0
6                                   Interviewer number             0
7                                     Day of interview             0
8                                   Month of interview             0
9                                    Year of interview             0
10                           Result of man's interview             0
11                                        Field editor             0
12                                    Data entry clerk             0
13                           Start of interview - Hour          1094
14                        Start of interview - Minutes          1094
15                             End of interview - Hour          1094
16                          End of interview - Minutes          1094
17                               Month of birth of man          1094
18                                Year of birth of man          1094
19                                          Age of man          1094
20                                Ever attended school          1094
21                    Highest level of school attended          1163
22               Highest grade completed at that level          1164
23                       Can read part of the sentence          6923
24          Frequency of reading newspaper or magazine          1731
25                 Frequency of listening to the radio          1094
26                            Frequency of watching TV          1094
27                                Ever used a computer          1094
28                Computer usage in the last 12 months          6763
29       Frequency of computer usage in the last month          7217
30                                  Ever used internet          1094
31                Internet usage in the last 12 months          6320
32       Frequency of Internet usage in the past month          6564
33              Ever used a mobile or non-mobile phone          1094
34   Mobile or non-mobile phone usage in the last 1...          1620
35   Frequency of mobile or non-mobile usage in the...          1850
36           Ever fathered any children with any women          1094
37                       Age when first child was born          4668
38               Any sons or daughters living with you          4668
39                                Sons living with you          5581
40                           Daughters living with you          5581
41           Any sons or daughters not living with you          4668
42                               Sons living elsewhere          6738
43                          Daughters living elsewhere          6738
44                       Ever had child who later died          1094
45                                           Boys dead          8126
46                                          Girls dead          8126
47                                  Children ever born          1094
48                              Same biological mother          4657
49   Number of women respondent fathered children with          7849
50                                 Month of last birth          4657
51                                  Year of last birth          4657
52   If she goes out with out telling husband: wife...          1094
53   If she neglects the children: wife beating jus...          1094
54   If she argues with husband: wife beating justi...          1094
55   If she refuses sex with husband: wife beating ...          1094
56       If she burns the food: wife beating justified          1094
57              Currently married or living with 

In [23]:
missing.describe()

missing vals
count    150.000000
mean    3868.420000
std     3233.806999
min        0.000000
25%     1094.000000
50%     2931.000000
75%     6782.000000
max     9008.000000

Drop columns with over 7000 missing values out of 9008.

In [24]:
drop_col = missing[missing["missing vals"] >7000]
drop_col

col  missing vals
29       Frequency of computer usage in the last month          7217
45                                           Boys dead          8126
46                                          Girls dead          8126
49   Number of women respondent fathered children with          7849
59     How many other wives / partners living with you          8822
61                                      Marital status          8625
65                                  Age at first union          8964
73               Condom used with prior sexual partner          8157
74                Relationship to prior sexual partner          8157
75        Sex with any other man in the last 12 months          8157
76            Number of sex partners in last 12 months          8793
97                                 Age at circumcision          8154
98                      Person performing circumcision          8154
99                               Place of circumcision          8154
103   Number of cigarettes smoked in the last 24 hours          7533
104  Number of days when cigarettes were smoked in ...          7533
106  Used any smoked tobacco products during the la...          7728
107             Type of smoked tobacco product: Cigars          8980
108         Type of smoked tobacco product: Water pipe          9005
109         Type of smoked tobacco product: Cigarillos          9006
110               Type of smoked tobacco product: Pipe          9003
111              Type of smoked tobacco product: Other          8998
112  Number of days when tobacco products where smo...          8293
114  Used any smokeless tobacco products during the...          8543
115  Type of smokeless tobacco product used: Chewin...          9008
116      Type of smokeless tobacco product used: Snuff          8882
117        Type of smokeless tobacco product used: Dip          9005
118      Type of smokeless tobacco product used: Other          9008
119  Number of days when smokeless tobacco products...          8879
128      Net treated with an insecticide when obtained          8649
129                Net soaked or dipped since obtained          8569
130                    Months ago net soaked or dipped          8964
134                       Person 3 who slept under net          7273
135                       Person 4 who slept under net          8109
143                  Date of birth of last child (CMC)          9008

In [25]:
#List of columns with too many missing values

to_drop = list(drop_col["col"])
to_drop

['Frequency of computer usage in the last month',
 'Boys dead',
 'Girls dead',
 'Number of women respondent fathered children with',
 'How many other wives / partners living with you',
 'Marital status',
 'Age at first union',
 'Condom used with prior sexual partner',
 'Relationship to prior sexual partner',
 'Sex with any other man in the last 12 months',
 'Number of sex partners in last 12 months',
 'Age at circumcision',
 'Person performing circumcision',
 'Place of circumcision',
 'Number of cigarettes smoked in the last 24 hours',
 'Number of days when cigarettes were smoked in past month',
 'Used any smoked tobacco products during the last month',
 'Type of smoked tobacco product: Cigars',
 'Type of smoked tobacco product: Water pipe',
 'Type of smoked tobacco product: Cigarillos',
 'Type of smoked tobacco product: Pipe',
 'Type of smoked tobacco product: Other',
 'Number of days when tobacco products where smoked in past month',
 'Used any smokeless tobacco products during the l

In [26]:
#Dropping the columns with too many missing values

raw2 = raw.drop(to_drop, axis = 1)

In [27]:
raw2.head(20)

Cluster number  Household number  Line number  Cluster number  \
0                1                17            1               1   
1                1                20            1               1   
2                2                 1            1               2   
3                2                 1            5               2   
4                2                 1            8               2   
5                2                 5            1               2   
6                2                 7            1               2   
7                2                10            1               2   
8                2                11            1               2   
9                2                13            1               2   
10               2                14            2               2   
11               2                14            3               2   
12               2                14            5               2   
13               2                17            1               2   
14               2                19            1               2   
15               2                20            1               2   
16               2                22            1               2   
17               2                25            1               2   
18               3                 3            1               3   
19               3                 4            1               3   

    Household number  Man's line number  Interviewer number  Day of interview  \
0                 17                  1                  14                 7   
1                 20                  1                  14                 7   
2                  1                  1                   9                 8   
3                  1                  5                   9                12   
4                  1                  8                   9                 8   
5                  5                  1                   9                 8   
6                  7                  1                   9                12   
7                 10                  1                   6                12   
8                 11                  1                  12                 7   
9                 13                  1                  12                 7   
10                14                  2                  12                13   
11                14                  3                  12                15   
12                14                  5                  12                 7   
13                17                  1                  10                13   
14                19                  1                  10                13   
15                20                  1                  10                12   
16                22                  1                   8                12   
17                25                  1                   8                 7   
18                 3                  1                   5                16   
19                 4                  1                   5                11   

    Month of interview  Year of interview Result of man's interview  \
0                    4               2014                 Completed   
1                    4               2014                 Completed   
2                    4               2014                 Completed   
3                    4               2014               Not at home   
4                    4               2014                 Completed   
5                    4               2014               Not at home   
6                    4               2014                 Completed   
7                    4               2014                 Completed   
8                    4               2014                 Completed   
9                    4               2014                 Completed   
10                   4               2014                 Completed   


# Check dataypes for all columns

In [28]:
raw2.dtypes

Cluster number                                                                      int64
Household number                                                                    int64
Line number                                                                         int64
Cluster number                                                                      int64
Household number                                                                    int64
Man's line number                                                                   int64
Interviewer number                                                                  int64
Day of interview                                                                    int64
Month of interview                                                                  int64
Year of interview                                                                   int64
Result of man's interview                                                          object
Field edit

- Most float64 can convert to int64. Except for 'Wealth index score'
- 'Date of interview women (CMC)', 'Date of birth of woman (CMC)', 'Date of marriage (CMC)' values do not really make sense so we will drop them.
- 'Children ever born' is duplicated 3 times.

In [29]:
# Dropping columns that do not add value

raw2 = raw2.drop(['Date of interview women (CMC)', 'Date of birth of woman (CMC)', 'Date of marriage (CMC)'], axis = 1)


In [30]:
float_col = raw2.select_dtypes('float64').columns.to_list()
float_col = float_col[:-1]
float_col

['Start of interview - Hour',
 'Start of interview - Minutes',
 'End of interview - Hour',
 'End of interview - Minutes',
 'Month of birth of man',
 'Year of birth of man',
 'Age of man',
 'Highest grade completed at that level',
 'Age when first child was born',
 'Sons living with you',
 'Daughters living with you',
 'Sons living elsewhere',
 'Daughters living elsewhere',
 'Children ever born',
 'Month of last birth',
 'Year of last birth',
 'Month of first union',
 'Year of first union',
 'Age at first sexual intercourse',
 'Time since last sexual intercourse (number)',
 'Number of sex partners in lifetime',
 'Age when cigarette was smoked for the first time',
 'Age when alcohol was used for the first time',
 'Number of days when at least one drink of alcohol was used in past month',
 'Number of drinks usually consumed',
 'Net number',
 'Months ago net obtained',
 'Person 1 who slept under net',
 'Person 2 who slept under net',
 'Age at first marriage/union',
 'Children ever born',
 

In [31]:
raw2[float_col].head()

Start of interview - Hour  Start of interview - Minutes  \
0                       17.0                          59.0   
1                       17.0                          32.0   
2                       10.0                          37.0   
3                        NaN                           NaN   
4                       10.0                          53.0   

   End of interview - Hour  End of interview - Minutes  Month of birth of man  \
0                     18.0                         7.0                    5.0   
1                     17.0                        42.0                    5.0   
2                     10.0                        52.0                    2.0   
3                      NaN                         NaN                    NaN   
4                     11.0                        10.0                    2.0   

   Year of birth of man  Age of man  Highest grade completed at that level  \
0                1984.0        29.0                                   31.0   
1                1976.0        37.0                                   31.0   
2                1973.0        41.0                                   17.0   
3                   NaN         NaN                                    NaN   
4                1993.0        21.0                                   24.0   

   Age when first child was born  Sons living with you  \
0                            NaN                   NaN   
1                            NaN                   NaN   
2                           21.0                   1.0   
3                            NaN                   NaN   
4                            NaN                   NaN   

   Daughters living with you  Sons living elsewhere  \
0                        NaN                    NaN   
1                        NaN                    NaN   
2                        1.0                    1.0   
3                        NaN                    NaN   
4                        NaN                    NaN   

   Daughters living elsewhere  Children ever born  Children ever born  \
0                         NaN                 0.0                 0.0   
1                         NaN                 0.0                 0.0   
2                         0.0                 3.0                 3.0   
3                         NaN                 NaN                 NaN   
4                         NaN                 0.0                 0.0   

   Month of last birth  Year of last birth  Month of first union  \
0                  NaN                 NaN                   9.0   
1                  NaN                 NaN                   2.0   
2                  5.0              2012.0                   8.0   
3                  NaN                 NaN                   NaN   
4                  NaN                 NaN                   NaN   

   Year of first union  Age at first sexual intercourse  \
0               2013.0                             20.0   
1               2014.0                             37.0   
2               1991.0                             20.0   
3                  NaN                              NaN   
4                  NaN                             19.0   

   Time since last sexual intercourse (number)  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          NaN   
4                                          7.0   

   Number of sex partners in lifetime  \
0                                 5.0   
1                                 1.0   
2                                 5.0   
3                                 NaN   
4                                 3.0   

   Age when cigarette was smoked for the first time  \
0                                               NaN   
1                                               NaN   
2                                              18.0   
3                      

In [32]:
# Remove na values and replace with median

for x in float_col:
    raw2[x] = raw2[x].fillna(raw2[x].median())


In [33]:
raw2[float_col].isna().sum()

Start of interview - Hour                                                      0
Start of interview - Minutes                                                   0
End of interview - Hour                                                        0
End of interview - Minutes                                                     0
Month of birth of man                                                          0
Year of birth of man                                                           0
Age of man                                                                     0
Highest grade completed at that level                                          0
Age when first child was born                                                  0
Sons living with you                                                           0
Daughters living with you                                                      0
Sons living elsewhere                                                          0
Daughters living elsewhere  

In [34]:
# Drop duplicate columns

raw2 = raw2.drop(["Children ever born"], axis = 1)

In [35]:
raw2["Children ever born"] = raw2["Children surviving"] + raw2["Children dead"]
raw2["Children ever born"] = raw2["Children ever born"].astype(int)

In [36]:
# Convert columns from float64 to int64 
for x in float_col:
    raw2[x] = raw2[x].astype(int)

In [37]:
raw2.head()

Cluster number  Household number  Line number  Cluster number  \
0               1                17            1               1   
1               1                20            1               1   
2               2                 1            1               2   
3               2                 1            5               2   
4               2                 1            8               2   

   Household number  Man's line number  Interviewer number  Day of interview  \
0                17                  1                  14                 7   
1                20                  1                  14                 7   
2                 1                  1                   9                 8   
3                 1                  5                   9                12   
4                 1                  8                   9                 8   

   Month of interview  Year of interview Result of man's interview  \
0                   4               2014                 Completed   
1                   4               2014                 Completed   
2                   4               2014                 Completed   
3                   4               2014               Not at home   
4                   4               2014                 Completed   

   Field editor  Data entry clerk  Start of interview - Hour  \
0             2                20                         17   
1             2                20                         17   
2             1                40                         10   
3             1                40                         12   
4             1                40                         10   

   Start of interview - Minutes  End of interview - Hour  \
0                            59                       18   
1                            32                       17   
2                            37                       10   
3                            29                       12   
4                            53                       11   

   End of interview - Minutes  Month of birth of man  Year of birth of man  \
0                           7                      5                  1984   
1                          42                      5                  1976   
2                          52                      2                  1973   
3                          29                      6                  1986   
4                          10                      2                  1993   

   Age of man Ever attended school Highest level of school attended  \
0          29                  Yes                           Higher   
1          37                  Yes                           Higher   
2          41                  Yes                          Primary   
3          28                  NaN                              NaN   
4          21                  Yes                        Secondary   

   Highest grade completed at that level Can read part of the sentence  \
0                                     31                           NaN   
1                                     31                           NaN   
2                                     17   Able to read whole sentence   
3                                     23                           NaN   
4                                     24                           NaN   

  Frequency of reading newspaper or magazine  \
0                           Almost every day   
1                       At least once a week   
2                                 Not at all   
3                                        NaN   
4                      Less than once a week   

  Frequency of listening to the radio Frequency of watching TV  \
0                At least once a week    Less than once a week   
1                          Not at all         Almost every day   
2                    Almost every day    Less than once a week   
3                                 NaN               

In [38]:
raw2.dtypes

Cluster number                                                                      int64
Household number                                                                    int64
Line number                                                                         int64
Cluster number                                                                      int64
Household number                                                                    int64
Man's line number                                                                   int64
Interviewer number                                                                  int64
Day of interview                                                                    int64
Month of interview                                                                  int64
Year of interview                                                                   int64
Result of man's interview                                                          object
Field edit

In [39]:
# Check for duplicate rows
raw2.duplicated().sum()

0

# Examine the columns to look for incorrect values

In [40]:
# Check year born values
raw2["Year of birth of man"].describe()

count    9008.000000
mean     1985.307504
std        85.022613
min      1959.000000
25%      1978.000000
50%      1986.000000
75%      1993.000000
max      9999.000000
Name: Year of birth of man, dtype: float64

In [41]:
# Birth year of 9999 is incorrect

raw2[raw2["Year of birth of man"] >2023]

Cluster number  Household number  Line number  Cluster number  \
5874             451                21            1             451   

      Household number  Man's line number  Interviewer number  \
5874                21                  1                 716   

      Day of interview  Month of interview  Year of interview  \
5874                31                   3               2014   

     Result of man's interview  Field editor  Data entry clerk  \
5874                 Completed           701                37   

      Start of interview - Hour  Start of interview - Minutes  \
5874                         11                            23   

      End of interview - Hour  End of interview - Minutes  \
5874                       11                          39   

      Month of birth of man  Year of birth of man  Age of man  \
5874                     99                  9999          46   

     Ever attended school Highest level of school attended  \
5874                  Yes                        Secondary   

      Highest grade completed at that level Can read part of the sentence  \
5874                                     22                           NaN   

     Frequency of reading newspaper or magazine  \
5874                                 Not at all   

     Frequency of listening to the radio Frequency of watching TV  \
5874               Less than once a week    Less than once a week   

     Ever used a computer Computer usage in the last 12 months  \
5874                   No                                  NaN   

     Ever used internet Internet usage in the last 12 months  \
5874                 No                                  NaN   

     Frequency of Internet usage in the past month  \
5874                                           NaN   

     Ever used a mobile or non-mobile phone  \
5874                                    Yes   

     Mobile or non-mobile phone usage in the last 12 months  \
5874                                                Yes       

     Frequency of mobile or non-mobile usage in the past month  \
5874                              Less than once a week          

     Ever fathered any children with any women  Age when first child was born  \
5874                                       Yes                             24   

     Any sons or daughters living with you  Sons living with you  \
5874                                   Yes                     1   

      Daughters living with you Any sons or daughters not living with you  \
5874                          1                                        No   

      Sons living elsewhere  Daughters living elsewhere  \
5874                      1                           1   

     Ever had child who later died Same biological mother  \
5874                           Yes                    Yes   

      Month of last birth  Year of last birth  \
5874                    6                2006   

     If she goes out with out telling husband: wife beating justified  \
5874                                                 No                 

     If she neglects the children: wife beating justified  \
5874                                                 No     

     If she argues with husband: wife beating justified  \
5874                                                 No   

     If she refuses sex with husband: wife beating justified  \
5874                                                 No        

     If she burns the food: wife beating justified  \
5874                                            No   

     Currently married or living with a man  \
5874                 Yes, currently married   

     Haveing other wives or liveing with other women as if you are married  \
5874                                      No (Only one)                      

     Ever married or lived with a woman  \
5874                                NaN   

     Married or lived with a woman once or more than once  \
5874        

In [42]:
# Remove row 5874

raw2 = raw2[raw2["Year of birth of man"] < 2023]

In [43]:
# Check number of children values
raw2["Children ever born"].describe()

count    9007.000000
mean        1.679472
std         2.177083
min         0.000000
25%         0.000000
50%         1.000000
75%         3.000000
max        30.000000
Name: Children ever born, dtype: float64

In [44]:
# Anything above 20 may be worth looking into but we will leave these rows in for now

raw2[raw2["Children ever born"] >25]

Cluster number  Household number  Line number  Cluster number  \
2504             192                17            1             192   
3325             259                21            1             259   
3875             298                13            1             298   

      Household number  Man's line number  Interviewer number  \
2504                17                  1                 215   
3325                21                  1                 312   
3875                13                  1                 413   

      Day of interview  Month of interview  Year of interview  \
2504                31                   3               2014   
3325                28                   3               2014   
3875                12                   3               2014   

     Result of man's interview  Field editor  Data entry clerk  \
2504                 Completed           202                22   
3325                 Completed           305                17   
3875                 Completed           401                 6   

      Start of interview - Hour  Start of interview - Minutes  \
2504                         11                            34   
3325                         14                             3   
3875                         10                             8   

      End of interview - Hour  End of interview - Minutes  \
2504                       11                          51   
3325                       14                          17   
3875                       10                          32   

      Month of birth of man  Year of birth of man  Age of man  \
2504                      3                  1969          45   
3325                      6                  1966          47   
3875                      9                  1960          53   

     Ever attended school Highest level of school attended  \
2504                  Yes                        Secondary   
3325                  Yes                          Primary   
3875                  Yes                          Primary   

      Highest grade completed at that level Can read part of the sentence  \
2504                                     21                           NaN   
3325                                     16   Able to read whole sentence   
3875                                     12   Able to read whole sentence   

     Frequency of reading newspaper or magazine  \
2504                      Less than once a week   
3325                                 Not at all   
3875                                 Not at all   

     Frequency of listening to the radio Frequency of watching TV  \
2504                    Almost every day               Not at all   
3325                          Not at all    Less than once a week   
3875                          Not at all    Less than once a week   

     Ever used a computer Computer usage in the last 12 months  \
2504                   No                                  NaN   
3325                   No                                  NaN   
3875                   No                                  NaN   

     Ever used internet Internet usage in the last 12 months  \
2504                 No                                  NaN   
3325                 No                                  NaN   
3875                 No                                  NaN   

     Frequency of Internet usage in the past month  \
2504                                           NaN   
3325                                           NaN   
3875                                           NaN   

     Ever used a mobile or non-mobile phone  \
2504                                    Yes   
3325                                    Yes   
3875                                    Yes   

     Mobile or non-mobile phone usage in the last 12 months  \
2504                                                Yes       
3325                                                Yes       
3875      

In [45]:
# Check age at first sexual intercourse

raw2["Age at first sexual intercourse"].describe()

count    9007.000000
mean       17.607639
std        15.771067
min         0.000000
25%        15.000000
50%        18.000000
75%        20.000000
max        99.000000
Name: Age at first sexual intercourse, dtype: float64

In [46]:
raw2["Age at first sexual intercourse"].value_counts()

18    1893
0     1837
20     794
19     623
17     606
16     479
22     403
21     371
15     341
25     274
23     268
95     257
24     233
14     128
26     112
27      69
28      65
13      59
30      41
29      38
12      34
10      21
32      14
11       7
8        6
31       5
36       5
99       4
9        4
35       3
39       2
34       2
37       2
41       2
33       2
47       1
40       1
38       1
Name: Age at first sexual intercourse, dtype: int64

In [47]:
raw2[raw2["Age at first sexual intercourse"] > 50]

Cluster number  Household number  Line number  Cluster number  \
52                 4                23            1               4   
99                 7                23            1               7   
600               46                16            1              46   
637               50                 6            1              50   
639               50                14            1              50   
719               56                17            1              56   
722               57                 1            1              57   
730               57                14            1              57   
733               57                20            1              57   
743               58                14            1              58   
765               60                 3            2              60   
781               61                 8            1              61   
863               68                 9            1              68   
958               75                18            1              75   
1041              82                25            1              82   
1305             105                18            1             105   
1388             112                17            2             112   
1404             113                20            1             113   
1460             118                22            5             118   
1475             120                 6            1             120   
1509             122                10            1             122   
1510             122                13            1             122   
1519             123                12            1             123   
1715             138                19            2             138   
1741             140                 3            1             140   
1759             141                15            1             141   
2316             178                23            3             178   
2321             179                 8            1             179   
2434             187                 7            1             187   
2504             192                17            1             192   
2581             197                23            1             197   
2592             198                18            1             198   
2800             214                 7            1             214   
2837             217                 6            1             217   
2932             226                 9            3             226   
2945             227                13            2             227   
3076             239                11            1             239   
3077             239                12            1             239   
3219             250                24            1             250   
3361             262                19            1             262   
3490             272                 5            1             272   
3578             278                21            1             278   
3594             279                23            1             279   
3668             284                11            1             284   
3679             285                12            1             285   
3688             285                24            1             285   
3708             287                14            1             287   
3765             291                 8            1             291   
3767             291                12            3             291   
3800             293                 6            1             293   
3802             293                 9            2             293   
3803             293                10            1             293   
3819             294                13            1             294   
3824             294                22            1             294   
3888             299                11            1             299   
3998             306          

There are a lot of values stating the individual was 95 during their first intercourse. These values are incorrect but instead of dropping them, we should replace them with the median.

In [48]:
raw2["Age at first sexual intercourse"].median()

18.0

In [49]:
# Replace values where Age at first sexual intercourse is over 90 to be the median value
raw2.loc[raw2['Age at first sexual intercourse'] > 50] = raw2["Age at first sexual intercourse"].median()

# Replace values where Age at first sexual intercourse is 0 to be the mean value
raw2.loc[raw2['Age at first sexual intercourse'] == 0] = raw2["Age at first sexual intercourse"].mean()

In [50]:
# Check values for Time since last sexual intercourse
raw2["Time since last sexual intercourse (unit)"].value_counts()

Days ago              3044
15.374597535250361    1837
Months ago            1293
Weeks ago             1046
Year ago               428
18.0                   261
Special                  4
Name: Time since last sexual intercourse (unit), dtype: int64

In [51]:
# Replace 18 with Weeks ago
raw2.loc[raw2["Time since last sexual intercourse (unit)"] == 18.0] = "Weeks ago"

In [52]:
# Replace 15 with Weeks ago
raw2.loc[raw2["Time since last sexual intercourse (unit)"] == 15.374597535250361] = "Weeks ago"

In [53]:
# Check values for Time since last sexual intercourse
raw2["Time since last sexual intercourse (unit)"].value_counts()

Weeks ago     3144
Days ago      3044
Months ago    1293
Year ago       428
Special          4
Name: Time since last sexual intercourse (unit), dtype: int64

In [54]:
raw2.head()

Cluster number Household number Line number Cluster number Household number  \
0              1               17           1              1               17   
1              1               20           1              1               20   
2              2                1           1              2                1   
3              2                1           5              2                1   
4              2                1           8              2                1   

  Man's line number Interviewer number Day of interview Month of interview  \
0                 1                 14                7                  4   
1                 1                 14                7                  4   
2                 1                  9                8                  4   
3                 5                  9               12                  4   
4                 8                  9                8                  4   

  Year of interview Result of man's interview Field editor Data entry clerk  \
0              2014                 Completed            2               20   
1              2014                 Completed            2               20   
2              2014                 Completed            1               40   
3              2014               Not at home            1               40   
4              2014                 Completed            1               40   

  Start of interview - Hour Start of interview - Minutes  \
0                        17                           59   
1                        17                           32   
2                        10                           37   
3                        12                           29   
4                        10                           53   

  End of interview - Hour End of interview - Minutes Month of birth of man  \
0                      18                          7                     5   
1                      17                         42                     5   
2                      10                         52                     2   
3                      12                         29                     6   
4                      11                         10                     2   

  Year of birth of man Age of man Ever attended school  \
0                 1984         29                  Yes   
1                 1976         37                  Yes   
2                 1973         41                  Yes   
3                 1986         28                  NaN   
4                 1993         21                  Yes   

  Highest level of school attended Highest grade completed at that level  \
0                           Higher                                    31   
1                           Higher                                    31   
2                          Primary                                    17   
3                              NaN                                    23   
4                        Secondary                                    24   

  Can read part of the sentence Frequency of reading newspaper or magazine  \
0                           NaN                           Almost every day   
1                           NaN                       At least once a week   
2   Able to read whole sentence                                 Not at all   
3                           NaN                                        NaN   
4                           NaN                      Less than once a week   

  Frequency of listening to the radio Frequency of watching TV  \
0                At least once a week    Less than once a week   
1                          Not at all         Almost every day   
2                    Almost every day    Less than once a week   
3                                 NaN                      NaN   
4                At least once a week    Less than once a week   

  Ever used a computer Computer usage in the last 12 months  \
0     

In [55]:
raw2.shape

(9007, 111)

There are over 100 columns in this data set and needs a lot of tidying up. A better approach would be to fully remove columns that are not vital to the data question we are trying to analyze and focus on the columns left that need a proper clean.